cpp/src/linear_programming/cuopt_c.cu

```diff
@@ -197,10 +205,87 @@ cuopt_int_t cuOptCreateRangedProblem(cuopt_int_t num_constraints,
+    static_cast<problem_and_stream_view_t*>(problem);
+
+  // Create pdhg_t with proper initialization
+  pdhg_t* pdhg = new pdhg_t();
+
+  pdhg->problem_ptr = std::make_unique<detail::problem_t<cuopt_int_t, cuopt_float_t>>(*problem_and_stream_view->op_problem);
+  pdhg->pdhg_solver_ptr = create_pdhg_solver(*problem_and_stream_view->op_problem, *pdhg->problem_ptr);
+  *pdhg_ptr = static_cast<cuOptPDHG>(pdhg);
+  return CUOPT_SUCCESS;
+}
+
+cuopt_int_t cuOptGetPDHGDimensions(cuOptPDHG pdhg, cuopt_int_t* num_variables, cuopt_int_t* num_constraints)
+{
+  if (pdhg == nullptr) { return CUOPT_INVALID_ARGUMENT; }
+  pdhg_t* pdhg_ptr = static_cast<pdhg_t*>(pdhg);
+  *num_variables = pdhg_ptr->pdhg_solver_ptr->get_primal_solution().size();
+  *num_constraints = pdhg_ptr->pdhg_solver_ptr->get_dual_solution().size();
+  return CUOPT_SUCCESS;
+}
+
+cuopt_int_t cuOptGetPDHGDeviceIterate(cuOptPDHG pdhg, cuopt_float_t** x, cuopt_float_t** y)
+{
+  if (pdhg == nullptr) { return CUOPT_INVALID_ARGUMENT; }
+  pdhg_t* pdhg_ptr = static_cast<pdhg_t*>(pdhg);
+  *x = pdhg_ptr->pdhg_solver_ptr->get_primal_solution().data();
+  *y = pdhg_ptr->pdhg_solver_ptr->get_dual_solution().data();
+  return CUOPT_SUCCESS;
+}
+
+cuopt_int_t cuOptGetPDHGHostIterate(cuOptPDHG pdhg, cuopt_float_t* x,  cuopt_float_t* y, cuopt_float_t* x_prime, cuopt_float_t* y_prime)
+{
+  if (pdhg == nullptr) { return CUOPT_INVALID_ARGUMENT; }
+  pdhg_t* pdhg_ptr = static_cast<pdhg_t*>(pdhg);
+  pdhg_ptr->problem_ptr->handle_ptr->get_stream().synchronize();
+  raft::copy(x, pdhg_ptr->pdhg_solver_ptr->get_primal_solution().data(), pdhg_ptr->pdhg_solver_ptr->get_primal_solution().size(), pdhg_ptr->problem_ptr->handle_ptr->get_stream());
+  raft::copy(y, pdhg_ptr->pdhg_solver_ptr->get_dual_solution().data(), pdhg_ptr->pdhg_solver_ptr->get_dual_solution().size(), pdhg_ptr->problem_ptr->handle_ptr->get_stream());
+  raft::copy(x_prime, pdhg_ptr->pdhg_solver_ptr->get_potential_next_primal_solution().data(), pdhg_ptr->pdhg_solver_ptr->get_potential_next_primal_solution().size(), pdhg_ptr->problem_ptr->handle_ptr->get_stream());
+  raft::copy(y_prime, pdhg_ptr->pdhg_solver_ptr->get_potential_next_dual_solution().data(), pdhg_ptr->pdhg_solver_ptr->get_potential_next_dual_solution().size(), pdhg_ptr->problem_ptr->handle_ptr->get_stream());
+  pdhg_ptr->problem_ptr->handle_ptr->get_stream().synchronize();
+  return CUOPT_SUCCESS;
+}
+
+
+cuopt_int_t cuOptPDHGIterations(cuOptPDHG pdhg, cuopt_int_t num_iterations, cuopt_float_t* host_primal_step_size, cuopt_float_t* host_dual_step_size)
+{
+  if (pdhg == nullptr) { return CUOPT_INVALID_ARGUMENT; }
+  if (host_primal_step_size == nullptr) { return CUOPT_INVALID_ARGUMENT; }
+  if (host_dual_step_size == nullptr) { return CUOPT_INVALID_ARGUMENT; }
+  pdhg_t* pdhg_ptr = static_cast<pdhg_t*>(pdhg);
+
+  rmm::device_scalar<cuopt_float_t> primal_step_size(*host_primal_step_size, pdhg_ptr->problem_ptr->handle_ptr->get_stream());
```

Comment:
If rmm is not allocated in pool mode, having to allocate a scalar during each PDHG iteration will severally hurt performance